[AMDGPU] Match bitsin(typeof(x)) - ctpop(x) to s_bcnt0_i32

linuxrocks123 opened the pull request:
This PR optimizes the pattern bitsin(typeof(x)) - popcnt(x) to s_bcnt0_i32 on AMDGPU.  It also creates a Clang builtin for s_bcnt0_i32 so that users can call this instruction directly instead of relying on the compiler to match this pattern.

linuxrocks123 committed:
Initial work

linuxrocks123 committed:
Update testcases

LU-JOHN commented on llvm/lib/Target/AMDGPU/AMDGPUCodeGenPrepare.cpp:
```diff
@@ -93,6 +95,13 @@ static cl::opt<bool> DisableFDivExpand(
   cl::ReallyHidden,
   cl::init(false));
 
+// Disable processing of fdiv so we can better test the backend implementations.
```
Comment needs to be updated.

LU-JOHN commented on llvm/include/llvm/IR/IntrinsicsAMDGPU.td:
```diff
@@ -2359,6 +2359,14 @@ def int_amdgcn_mbcnt_hi :
             [IntrNoMem]>;
 
+def int_amdgcn_bcnt032_lo :
+  ClangBuiltin<"__builtin_amdgcn_bcnt032_lo">,
```
Is "bcnt032_lo" the name we want to use?  For comparison:

```
  ClangBuiltin<"__builtin_amdgcn_sad_u8">,
  ClangBuiltin<"__builtin_amdgcn_msad_u8">,
```

follows the mnemonic without an initial "v_".

linuxrocks123 replied:
@LU-JOHN I removed the 0, which I feel makes them close to the existing `__builtin_amdgcn_mbcnt_lo`.  What do you think?

Re the option, I think most of the other transformations have options, so I feel this one should as well.

linuxrocks123 committed:
Don't perform optimization on vector types

linuxrocks123 committed:
Review changes

arsenm reviewed (changes requested):
This should not introduce a builtin or intrinsic. This can be purely done in tablegen patterns without the intermediate step

jmmartinez commented on llvm/lib/Target/AMDGPU/AMDGPUCodeGenPrepare.cpp:
```diff
@@ -1977,6 +1989,37 @@ Value *AMDGPUCodeGenPrepareImpl::applyFractPat(IRBuilder<> &Builder,
+bool AMDGPUCodeGenPrepareImpl::visitCtpop(IntrinsicInst &I) {
+  uint32_t BitWidth, DestinationWidth, IntrinsicWidth;
+  if (!I.hasOneUse() || !I.getType()->isIntegerTy() ||
+      !ST.hasBCNT(BitWidth = I.getType()->getIntegerBitWidth()))
```
Please do not try to put everything into a one-liner.

There is no need to assign variables at the same time you're trying to pass an argument to the function. 
It makes the code shorter, but harder to read.

Doing one single thing at a time makes everything much more readable.


```suggestion
  uint32_t DestinationWidth, IntrinsicWidth;
  
  if(!I.hasOneUse())
    return false;
    
  IntegerType *IType = dyn_cast<IntegerType>(I.getType());
  if(!IType)
    return false;
    
  uint32_t BitWidth = IType->getIntegerBitWidth();
  if (!ST.hasBCNT(BitWidth))
    return false;
```

In the if below, I was really not expecting the assignments on each side of the comparison either:
```cpp
  if ((DestinationWidth = MustBeSub->getType()->getIntegerBitWidth()) !=
      (IntrinsicWidth = TransformedIns->getType()->getIntegerBitWidth()))
```

jmmartinez commented on llvm/lib/Target/AMDGPU/AMDGPUCodeGenPrepare.cpp:
```diff
@@ -1977,6 +1989,37 @@ Value *AMDGPUCodeGenPrepareImpl::applyFractPat(IRBuilder<> &Builder,
+
+  if ((DestinationWidth = MustBeSub->getType()->getIntegerBitWidth()) !=
+      (IntrinsicWidth = TransformedIns->getType()->getIntegerBitWidth()))
+    TransformedIns = cast<Instruction>(Builder.CreateZExtOrTrunc(
```
If you turn `TransformedIns` into a `Value*` you would avoid the `cast<Instruction>`.

linuxrocks123 commented on clang/include/clang/Basic/BuiltinsAMDGPU.def:
```diff
@@ -63,6 +63,9 @@ BUILTIN(__builtin_amdgcn_grid_size_z, "Ui", "nc")
 BUILTIN(__builtin_amdgcn_mbcnt_lo, "UiUiUi", "nc")
 
+BUILTIN(__builtin_amdgcn_bcnt32_lo, "UiUi", "nc")
+BUILTIN(__builtin_amdgcn_bcnt64_lo, "UiWUi", "nc")
```
k, removed it

linuxrocks123 committed:
Review changes:
 - Add tests
 - Remove builtin (users will need inline assembly if pattern match fails)

arsenm commented on llvm/include/llvm/IR/IntrinsicsAMDGPU.td:
```diff
@@ -2359,6 +2359,12 @@ def int_amdgcn_mbcnt_hi :
   DefaultAttrsIntrinsic<[llvm_i32_ty], [llvm_i32_ty, llvm_i32_ty],
             [IntrNoMem]>;
 
+def int_amdgcn_bcnt32_lo :
```
This does not need an intrinsic. This doesn't help match the pattern in any cases, and introduces new support burdens to every optimization

arsenm reviewed (changes requested):
This should not introduce a new intrinsic and only needs a tablegen pattern. The one benefit you get out doing this fold in the IR would be sinking a popcnt out of block, but you can do that just by handling this case in isProfitableToSinkOperands, and that avoids teaching all of the known bits / sign bits / simplify demanded bits about this

linuxrocks123 committed:
Reviewer-suggested refactoring

jmmartinez commented on llvm/lib/Target/AMDGPU/AMDGPUCodeGenPrepare.cpp:
```diff
@@ -93,6 +96,13 @@ static cl::opt<bool> DisableFDivExpand(
+                 cl::desc("Prevent transforming bitsin(typeof(x)) - "
+                          "popcount(x) to bcnt0(x) in AMDGPUCodeGenPrepare"),
+                 cl::ReallyHidden, cl::init(false));
+
```
You can remove this. It is not used anywhere.

linuxrocks123 committed:
Revert implementation

linuxrocks123 committed:
Use S-expressions instead

jayfoad commented on llvm/test/CodeGen/AMDGPU/s_cmp_0.ll:
```diff
@@ -465,17 +463,16 @@ define amdgpu_ps i32 @bcnt032(i32 inreg %val0) {
-; CHECK-NEXT:    ; return to shader part epilog
+; CHECK-NEXT: s_bcnt0_i32_b64 s0, s[0:1]
+; CHECK-NEXT: s_mov_b32 s1, 0
+; CHECK-NEXT: s_cmp_lg_u64 s[0:1], 0
```
As a follow up, it would be good if we could somehow generate a 32-bit compare instead here:
```suggestion
; CHECK-NEXT: s_cmp_lg_u32 s0, 0
```

jayfoad reviewed (commented):
Looks good overall.

linuxrocks123 commented on llvm/lib/Target/AMDGPU/SOPInstructions.td:
```diff
@@ -1884,6 +1886,13 @@ def : GCNPat <
 >;
 
+def : GCNPat <
+   (i64 (UniformBinFrag<sub> 64, (UniformUnaryFrag<ctpop> i64:$src))),
```
Done

linuxrocks123 commented on llvm/lib/Target/AMDGPU/SOPInstructions.td:
```diff
@@ -264,7 +264,9 @@ def S_BREV_B64 : SOP1_64 <"s_brev_b64",
 let Defs = [SCC] in {
-def S_BCNT0_I32_B32 : SOP1_32 <"s_bcnt0_i32_b32">;
+def S_BCNT0_I32_B32 : SOP1_32 <"s_bcnt0_i32_b32",
+  [(set i32:$sdst, (UniformBinFrag<sub> 32, (UniformUnaryFrag<ctpop> i32:$src0)))]
```
Done

linuxrocks123 committed:
Review changes

linuxrocks123 committed:
Newline

LU-JOHN reviewed (approved):
LGTM

jayfoad commented on llvm/test/CodeGen/AMDGPU/s_bcnt0.ll:
```diff
@@ -0,0 +1,110 @@
+  %cmp = icmp ne i64 %result2, 0
+  %zext = zext i1 %cmp to i32
+  ret i32 %zext
+}
```
Missing newline at end of file

linuxrocks123 replied:
Fixed

linuxrocks123 committed:
Fix testcases

arsenm reviewed (commented):
Description is out of date

arsenm commented on llvm/test/CodeGen/AMDGPU/s_bcnt0.ll:
```diff
@@ -0,0 +1,110 @@
+; CHECK-NEXT:    v_subb_co_u32_e64 v6, s[0:1], 0, 0, vcc
+; CHECK-NEXT:    ;;#ASMSTART
+; CHECK-NEXT:    ; use v[5:6]
+; CHECK-NEXT:    ;;#ASMEND
```
Oh this is bad bug. Your SGPR constraint was lost and silently transmuted into a VGPR. Not related to this PR though,  for your purposes you're just using an overly complicated test.

linuxrocks123 replied:
Yikes!  Do you want to file the issue or should I?

linuxrocks123 commented on llvm/lib/Target/AMDGPU/SOPInstructions.td:
```diff
@@ -1884,6 +1886,13 @@ def : GCNPat <
+def : GCNPat <
+   (i64 (UniformBinFrag<sub> 64, (ctpop i64:$src))),
+    (i64 (REG_SEQUENCE SReg_64,
+     (i32 (COPY_TO_REGCLASS (S_BCNT0_I32_B64 $src), SReg_32)), sub0,
```
@arsenm I have no clue.  I put it there because I was aping similar code in that file that seemed to use it when copying from 32-bit to 64-bit.  I can remove it and see if it still works?

linuxrocks123 committed:
Change testcases

arsenm commented on llvm/test/CodeGen/AMDGPU/s_bcnt0.ll:
```diff
@@ -0,0 +1,96 @@
+  %val0 = trunc i64 %val to i32
+  %result = call i32 @llvm.ctpop.i32(i32 %val0)
+  %result2 = sub i32 32, %result
+  call void asm "; use $0", "s"(i32 %result2)
```
This is still hitting that asm bug, but not sure why you're using the asm here in the first place. You just need a VGPR return?

linuxrocks123 commented:
@arsenm LGTM?

linuxrocks123 commented:
@arsenm just saw your testcase comments.  One sec.

linuxrocks123 committed:
Testcase updates (maybe revert this)

linuxrocks123 commented:
@arsenm okay, comments dealt with.  If you'd like to preserve the global isel error, I should revert or partially revert the last commit.  Otherwise, LGTM?

arsenm commented on llvm/test/CodeGen/AMDGPU/s_bcnt0.ll:
```diff
@@ -0,0 +1,110 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 6
+; RUN: llc -mtriple=amdgcn -mcpu=gfx900 < %s | FileCheck %s
```
This is failing on the broken v-to-s copy for the asm input. I think this is the global isel flavored version of the case the DAG path miscompiles into ignoring the SGPR constraint

arsenm commented on llvm/test/CodeGen/AMDGPU/s_bcnt0.ll:
```diff
@@ -0,0 +1,85 @@
+; CHECK-LABEL: bcnt032_ctpop_multiple_uses:
+; CHECK:       ; %bb.0:
+; CHECK-NEXT:    s_bcnt1_i32_b32 s1, s0
+; CHECK-NEXT:    s_bcnt0_i32_b32 s0, s0
```
I'm not sure we should do this in the multiple use case. It's not worse, but it's trading for an equivalently good instruction

arsenm commented on llvm/test/CodeGen/AMDGPU/s_bcnt0.ll:
```diff
@@ -0,0 +1,85 @@
+  ret i32 %zext
+}
+
+define i32 @bcnt064_not_for_vregs(i64 %val0) {
```
Not sure what "not_for_vregs" is supposed to mean, this case tests VGPR input + VGPR use result

linuxrocks123 replied:
Fixed

arsenm reviewed (commented):
LGTM except for test naming

linuxrocks123 committed:
Merge branch 'master' into swdev-562125

linuxrocks123 committed:
Rename confusingly named tests

linuxrocks123 committed:
Update testcase

jayfoad reviewed (approved):
Really the benefit still seems marginal, but I guess it's not doing any harm.

> If you have reason to believe that the other pattern will also match, I'll add it, too. Would you like me to do that?

Yes please but it could be a separate PR with its own test coverage.
